File: src/omnia/data/structure/tuple/ImmutableQuintuplet.java
package omnia.data.structure.tuple;

import java.util.Iterator;
import java.util.OptionalInt;
import java.util.stream.Stream;
import omnia.data.structure.immutable.ImmutableList;

class ImmutableQuintuplet<T> extends ImmutableQuintuple<T, T, T, T, T> implements Quintuplet<T> {

  ImmutableQuintuplet(T first, T second, T third, T fourth, T fifth) {
    super(first, second, third, fourth, fifth);
  }

  @Override
  public Quadruplet<T> dropFirst() {
    return Tuplet.of(second(), third(), fourth(), fifth());
  }

  @Override
  public Quadruplet<T> dropSecond() {
    return Tuplet.of(first(), third(), fourth(), fifth());
  }

  @Override
  public Quadruplet<T> dropThird() {
    return Tuplet.of(first(), second(), fourth(), fifth());
  }

  @Override
  public Quadruplet<T> dropFourth() {
    return Tuplet.of(first(), second(), third(), fifth());
  }

  @Override
  public Quadruplet<T> dropFifth() {
    return Tuplet.of(first(), second(), third(), fourth());
  }

  @Override
  public Sextuplet<T> concat(T object) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), object);
  }

  @Override
  public Septuplet<T> concat(Couple<T, T> other) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), other.first(), other.second());
  }

  @Override
  public Octuplet<T> concat(Triple<T, T, T> other) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), other.first(), other.second(), other.third());
  }

  @Override
  public Nonuplet<T> concat(Quadruple<T, T, T, T> other) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), other.first(), other.second(), other.third(), other.fourth());
  }

  @Override
  public Decuplet<T> concat(Quintuple<T, T, T, T, T> other) {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), other.first(), other.second(), other.third(), other.fourth(), other.fifth());
  }

  @Override
  public Iterator<T> iterator() {
    return toActualList().iterator();
  }

  @Override
  public T itemAt(int index) {
    return toActualList().itemAt(index);
  }

  @Override
  public OptionalInt indexOf(Object item) {
    return toActualList().indexOf(item);
  }

  @Override
  public Stream<T> stream() {
    return toActualList().stream();
  }

  @Override
  public boolean containsUnknownTyped(Object element) {
    return toActualList().containsUnknownTyped(element);
  }

  private ImmutableList<T> toActualList() {
    return ImmutableList.of(first(), second(), third(), fourth(), fifth());
  }
}

File: src/omnia/data/structure/tuple/ImmutableQuintuplet.kt
package omnia.data.structure.tuple

import java.util.OptionalInt
import java.util.function.Function
import java.util.stream.Stream
import omnia.data.structure.immutable.ImmutableList

internal class ImmutableQuintuplet<T>(first: T, second: T, third: T, fourth: T, fifth: T) :
  ImmutableQuintuple<T, T, T, T, T>(first, second, third, fourth, fifth), Quintuplet<T> {

  override fun <R> map(mapper: Function<in T, out R>): Quintuplet<R> {
    return Tuplet.of(
      mapper.apply(first()),
      mapper.apply(second()),
      mapper.apply(third()),
      mapper.apply(fourth()),
      mapper.apply(fifth())
    )
  }

  override fun dropFirst(): Quadruplet<T> {
    return Tuplet.of(second(), third(), fourth(), fifth())
  }

  override fun dropSecond(): Quadruplet<T> {
    return Tuplet.of(first(), third(), fourth(), fifth())
  }

  override fun dropThird(): Quadruplet<T> {
    return Tuplet.of(first(), second(), fourth(), fifth())
  }

  override fun dropFourth(): Quadruplet<T> {
    return Tuplet.of(first(), second(), third(), fifth())
  }

  override fun dropFifth(): Quadruplet<T> {
    return Tuplet.of(first(), second(), third(), fourth())
  }

  override fun concat(`object`: T): Sextuplet<T> {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), `object`)
  }

  override fun concat(other: Couple<T, T>): Septuplet<T> {
    return Tuplet.of(first(), second(), third(), fourth(), fifth(), other.first(), other.second())
  }

  override fun concat(other: Triple<T, T, T>): Octuplet<T> {
    return Tuplet.of(
      first(),
      second(),
      third(),
      fourth(),
      fifth(),
      other.first(),
      other.second(),
      other.third()
    )
  }

  override fun concat(other: Quadruple<T, T, T, T>): Nonuplet<T> {
    return Tuplet.of(
      first(),
      second(),
      third(),
      fourth(),
      fifth(),
      other.first(),
      other.second(),
      other.third(),
      other.fourth()
    )
  }

  override fun concat(other: Quintuple<T, T, T, T, T>): Decuplet<T> {
    return Tuplet.of(
      first(),
      second(),
      third(),
      fourth(),
      fifth(),
      other.first(),
      other.second(),
      other.third(),
      other.fourth(),
      other.fifth()
    )
  }

  override fun iterator(): Iterator<T> {
    return toActualList().iterator()
  }

  override fun itemAt(index: Int): T {
    return toActualList().itemAt(index)
  }

  override fun indexOf(item: Any?): OptionalInt {
    return toActualList().indexOf(item)
  }

  override fun stream(): Stream<T> {
    return toActualList().stream()
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return toActualList().containsUnknownTyped(item)
  }

  private fun toActualList(): ImmutableList<T> {
    return ImmutableList.of(first(), second(), third(), fourth(), fifth())
  }
}